plugin/cache: Add optional ristretto cache

huntharo opened the pull request:
## 1. Why is this pull request needed and what does it do?

- Option to use a cache with the following features:
  - Non-blocking adds to the cache (not making the requester of the non-cached item wait for the write lock to be acquired)
  - Key collision protection (ensure that a key collision will not return an answer to the wrong question)
  - Admission policy (ensure that single-request for an item does not push other items out of the cache)
  - LFU eviction instead of random (keep frequently and recently requested items in the cache)
- Future
  - Enable TTL eviction in the new cache
  - Eventually: make the new cache the default option if it is an improvement for most

### 2. Which issues (if any) are related?

#3695 

### 3. Which documentation changes (if any) need to be made?

- DONE: Move references to specific behaviors of the internal cache module to sections that apply only to that particular cache module (e.g. discussion of eviction being random)
- DONE: Document the new `ristretto` cache config option

### 4. Does this introduce a backward incompatible change or deprecation?

- Not at the moment - The new cache is added as a non-default option

### 5. To-Do

- [ ] Review and confirm metrics are still at least somewhat compatible with prior values
- [ ] Determine why the existing builtin cache got faster between the baseline and this PR
- [x] Review the Ristretto default hash and replace if slower than FNV64
   - Ristretto was using MemHash + xxHash for a 128 bit hash by default
   - xxHash alone was adding ~200 ns / op
   - Passed Ristretto.KeyToHash to revert to FNV64 hashes - nearly restored read performance to baseline

huntharo committed:
Add optional ristretto cache

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Add comments and unexport

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Update comment format

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Update plugin/cache/README.md

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Fix do integer to fix Sprintf warning

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Run cache tests against both types of caches

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Run cache tests for both types of cache backend

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Get plugin tests passing

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Go back to Name

QName has benefits and drawbacks... it makes the tests pass but is not testing the same thing

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Get the tests passing without QName

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Add additional benchmarks

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Add benchmarks, make tests pass without metrics

Signed-off-by: Harold <harold@pwrdrvr.com>

SuperQ commented on .vscode/launch.json:
```diff
@@ -0,0 +1,17 @@
+{
```
Please drop this file from the PR. We don't need `.vscode` stuff in the codebase.

huntharo replied:
Thanks for noticing.  I'm usually super clean about that... but this once slipped by.

huntharo committed:
Turn off prefetch in cache type tests

Prefetch was accidentally turned on and was preventing the conditions that were intended to be tested from being tested

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Fix parallel inserts read test

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Use unique random generators in each goroutine

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Add parallel read test

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Remove .vscode folder

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Ignore .vscode fodler

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Add a Balanced test

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Use the FNV hash with Ristretto

Signed-off-by: Harold <harold@pwrdrvr.com>

huntharo committed:
Add a rudimentary length stat

huntharo committed:
Add a .Len test - Fix Ristretto Len

- 8 byte alignment...

huntharo committed:
Add OnShutdown to cleanup Ristretto goroutines
- Ristretto creates goroutines to manage each cache
- We need to clean these up on reload

huntharo committed:
Fix comment for code review

huntharo committed:
Add TTL eviction option for Ristretto

huntharo committed:
Add setup test for ttlEvict

huntharo committed:
Document ttlEvict parameter
